data-structures/src/data/structure/linkedlist/LinkedList.java

```diff
@@ -0,0 +1,239 @@
+package data.structure.linkedlist;
+import static  java.util.Objects.*;
+
+public class LinkedList<E> {
+    private class Node {
+        E data;
+        Node next;
+        Node previous;
+
+        Node(final E data) {
+            this.data = data;
+            this.next = null;
+            this.previous = null;
+        }
+    }
+
+    private Node head;
+    private Node tail;
+    private int size;
+
+    public LinkedList() {
+        this.head = null;
+        this.tail = null;
+        size = 0;
+    }
+
+    public void add(E element) {
+        requireNonNull(element, "The key must not be null!");
+
+        if (isNull(head)) {
+            head = new Node(element);
+            tail = head;
+            size++;
+
+        } else {
+            tail.next = new Node(element);
+            final Node temp = tail;
+            tail = tail.next;
+            tail.previous = temp;
+            size++;
+        }
+    }
+
+    public void add(int index, E element) {
+        requireNonNull(element, "The element must not be null!");
+
+        if (index > size || index < 0) {
+            throw new IllegalArgumentException("Enter a proper index value!");
+        }

```

Comment:
This is used at multiple places, therefore is a good candidate for method extraction. Just like `requireNonNull`, this is a so-called guard function, checking the input. Thus, you could name the method something like `guardIndexIsValid(index)`.